Abacus error handling documentation (#1410)
kiesraad/abacus: documentatie/softwarearchitectuur/foutafhandeling.md

# Foutafhandeling in Abacus

In Abacus onderscheiden we de volgende soorten fouten:

- Connectie fouten
- Fatale fouten
- Functionele fouten

> - Authenticatie/authorizatie fouten

Reviewer: ```suggestion
- Authenticatie/autorisatie fouten
```

- Niet-gevonden fouten

> - Airgap fout

Reviewer: ```suggestion
- Air-gap fouten
```

## Connectie fouten

> Dit zijn fouten die optreden als de Abacus backend niet bereikbaar is, bijvoorbeeld door netwerk problemen, of omdat Abacus niet draait.

Reviewer: ```suggestion
Dit zijn fouten die optreden als de Abacus backend niet bereikbaar is, bijvoorbeeld door netwerkproblemen, of omdat Abacus niet draait.
```

> In dit geval toont Abacus een pop-up in het scherm met dat er een fout is opgetreden en het voorstel om het opnieuw te proberen.

Reviewer: ```suggestion
In dit geval toont Abacus middels een pop-up dat er geen verbinding meer is, met de mogelijkheid het opnieuw te proberen.
```

## Fatale fouten
Dit zijn over het algemeen programmeer-fouten, zoals een ongeldige JSON-structuur of een ongeldige actie tijdens het invoeren. 

Fatale fouten kunnen ook het resultaat zijn van het moedwillig invoeren van een ongeldige url.

Deze fouten worden afgehandeld middels een "Abacus is stuk" fout pagina.

## Functionele fouten
Functionele fouten zijn fouten die te verwachten zijn als gevolg van acties van de gebruiker, en die ook weer hersteld kunnen worden. Dit zijn bijvoorbeeld validatie fouten die door de backend teruggegeven worden.

> De afhandeling van deze fouten wordt gespecificeerd per use case.

Reviewer: ```suggestion
De afhandeling van deze fouten wordt gespecificeerd per use case en afgehandeld in de user interface van de betreffende functionaliteit.
```

> ## Authenticatie/authorizatie fouten

Reviewer: ```suggestion
## Authenticatie/autorisatie fouten
```

Wanneer de gebruiker niet (meer) is ingelogd, toont Abacus de mogelijkheid tot inloggen.

> Wanneer de gebruiker een pagina oproept die niet toegestaan is, toont Abacus een pagina die dat meldt.

Reviewer: ```suggestion
Wanneer de gebruiker een pagina oproept die vanuit diens rol niet toegestaan is, toont Abacus een pagina die dat meldt.
```

## Niet-gevonden fouten

Deze fout treedt op wanneer de gebruiker naar een url gaat die niet bestaat in Abacus, of naar een pagina gaat die hoort bij een entiteit die niet meer bestaat, zoals bijvoorbeeld een stembureau dat verwijderd is.

Beide fouten worden afgehandeld met een "De pagina is niet gevonden" pagina.

> ## Airgap fouten

Reviewer: ```suggestion
## Air-gap fouten
```

> Als Abacus detecteert dat het internet bereikbaar is, wordt er een passende foutmelding getoond.

Reviewer: ```suggestion
Als Abacus detecteert dat het internet bereikbaar is en Abacus dus niet air-gapped draait, wordt er een passende foutmelding getoond.
```

> Als Abacus detecteert dat het internet bereikbaar is, wordt er een passende foutmelding getoond.

Reviewer 2: Moeten we hier expliciet maken dat het de client-side airgap-detectie is? En niet de server-side detectie?

> Als Abacus detecteert dat het internet bereikbaar is, wordt er een passende foutmelding getoond.

Reviewer: Is dit er eentje voor de us cases @jschuurk-kr ? Wat willen we überhaupt doen met de air-gap detectie? En maken we onderscheid tussen server- en client? Behandelen we de hele installatie als 'besmet' op het moment dat er internetverbinding is geweest tijdens de invoerfase of daarna? Of kan je vrolijk doorwerken? Loggen we het ergens? Komt het in het proces-verbaal terecht? ('Op 19 maart tussen 12:07 en 13:37 was de telresultatensoftware tegen de afspraken in met internet verbonden')

> Als Abacus detecteert dat het internet bereikbaar is, wordt er een passende foutmelding getoond.

Reviewer: En over waar we de foutmelding tonen: wat mij betreft tonen we de melding op de server en op alle werkstations, en minimaal zolang de internetverbinding er is.

> Als Abacus detecteert dat het internet bereikbaar is, wordt er een passende foutmelding getoond.

Author: Wat mij betreft laten we dit document zo voor wat betreft air-gap omdat het inderdaad nog niet duidelijk of iig geïmplementeerd is

> Als Abacus detecteert dat het internet bereikbaar is, wordt er een passende foutmelding getoond.

Reviewer: Yes, eens. Voor dit document nu voldoende!
